dev-tools/ls-apis/src/cargo.rs

```diff
@@ -0,0 +1,384 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Extract API metadata from Cargo metadata
+
+use crate::ClientPackageName;
+use anyhow::bail;
+use anyhow::{anyhow, ensure, Context, Result};
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use cargo_metadata::Package;
+use cargo_metadata::{DependencyKind, PackageId};
+use std::collections::BTreeSet;
+use std::collections::{BTreeMap, VecDeque};
+
+/// Packages that may be flagged as clients because they directly depend on
+/// Progenitor, but which are not really clients.
+const IGNORED_NON_CLIENTS: &[&str] = &[
+    // omicron-common depends on progenitor so that it can define some generic
+    // error handling and a generic macro for defining clients.  omicron-common
+    // itself is not a progenitor-based client.
+    "omicron-common",
+    // propolis-mock-server uses Progenitor to generate types that are
+    // compatible with the real Propolis server.  It doesn't actually use the
+    // client and we don't really care about it for these purposes anyway.
+    "propolis-mock-server",
+];
+
+/// Query package and dependency-related information about a Cargo workspace
+pub struct Workspace {
```

Comment:
How do you feel about using my guppy library for this? We already use it within omicron and it does do a lot of this.

Reply:
Discussed offline.  This might have been a good idea, but seems not worth re-doing now.